### PR TITLE
Media Recording: Add Background Blur

### LINKED ISSUE #11909: Self Recording: Blur Background support
## Feature Description

Experiment with e.g. blurred background, masks/filters, using MediaPipe selfie segmentation and face mesh.

### LINKED ISSUE #12033: Media Recording: Broken image capture
## Bug Description

When trying to take a still picture in media recording mode, the photo is taken straight away instead of waiting for the countdown to finish.

## Expected Behaviour

The photo should only be taken after the countdown

## Screenshots

https://user-images.githubusercontent.com/841956/181281069-b89babde-f5aa-48d8-bf28-983c1c7e07b6.mov

## Additional Context

- Plugin Version: 1.24.0-alpha
- WordPress Version: 6.0.1
- Operating System: macOS
- Browser: Chrome

### LINKED ISSUE #12245: Self Recording blur background is not handling all cameras aspect ratio properly
## Bug Description
Originally reported by @spacedmonkey 
Mac OS - Chrome. Image is looked strange. ( before and after adding blur ). Makes me look thinner. Only does this on external webcam. 

image1

## Expected Behaviour

Camera image aspect ratio should be retained for all resolutions.

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### LINKED ISSUE #12246: Self Recording blur background causes image to freeze
## Bug Description
Originally reported by @spacedmonkey 
Mac OS - Safari, after setting blur background image freezes and never moves again. ( built in web cam and external web cam ). 

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### LINKED ISSUE #12249: Self Recording - Blur background not working as expected in Firefox
## Bug Description

Originally reported by @spacedmonkey 

Mac OS - Firefox. Open the webcam dialog, select blur background. Close dialog. Reopen, see weird glitchy version of my image.
## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

### PR DESCRIPTION
**Pull request by merapi**
## Summary

This version of Background Blur is based on the latest version of the [new Trimming flow PR](https://github.com/GoogleForCreators/web-stories-wp/pull/12022).
The Media Pipe object is very difficult to cooperate and it often throws errors while switching between no effect and blur effect. In this PR I managed to make it work consistently. The recording part is not yet merged into this version (two media recorders one used to get the camera feed, the second for canvas recording when the blur effect is active).

image

Edges are now smooth so it looks a lot better than the first version:
image

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [ ] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [ ] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [ ] I have added documentation where necessary
- [ ] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11909
Fixes #12033
Fixes #12245
Fixes #12249
Fixes #12246

### CODE REVIEW

File: packages/story-editor/package.json [hunk 1]
```diff
@@ -61,6 +61,7 @@
     "@googleforcreators/transform": "*",
     "@googleforcreators/units": "*",
     "@googleforcreators/url": "*",
+    "@mediapipe/selfie_segmentation": "^0.1.1632777926",
```
Comment (barklund) [inline packages/story-editor/package.json:65]: That's quite the version number! 😳

File: package-lock.json [hunk 2]
```diff
@@ -47468,7 +47476,7 @@
         "colorthief": "^2.3.2",
         "draft-js": "^0.11.7",
         "draft-js-export-html": "^1.4.1",
-        "draft-js-import-element": "git+ssh://git@github.com/swissspidy/draft-js-import-element.git",
+        "draft-js-import-element": "https://github.com/swissspidy/draft-js-import-element",
```
Comment (barklund) [inline package-lock.json]: Hmm, are these changes intended? Do they matter? Ping @swissspidy

**Review by barklund — COMMENTED** [review]
This works really well, but it's hard to review with the mixed code from multiple PR's. Let's have @felipebochehin87 quickly QA #11988, so we can merge #12022 and rebase this PR.

**Comment by barklund** [comment]
#12022 has been merged, so this can be updated to main to better reflect that actual changes.

File: package-lock.json (hunk 2 above)
Comment (swissspidy) [reply package-lock.json]: That usually happens when you configure Git to pull via HTTPS and push via SSH (and someone else hasn't configured it like that). Then a change like this to the lock file can happen. Doesn't really cause issues besides unnecessarily changing files.

File: packages/story-editor/package.json (hunk 1 above)
Comment (swissspidy) [reply packages/story-editor/package.json:65]: Timestamps :-)

**Comment by merapi** [comment]
Summary of the latest updates:
You will see two `useMediaRecorder` instances in the code: `mediaRecorder` and `canvasRecorder`.
When all effects are disabled (default state) we will do the recording in the old way (`mediaRecorder` will be the `currentRecorder`) - this is more performant way than always redrawing the camera stream onto the canvas and recording the canvas (with and without effects) - so it will be as snappy as possible on slower devices when you don't use any effects.

I also fixed https://github.com/GoogleForCreators/web-stories-wp/issues/12033

File: packages/story-editor/src/components/mediaRecording/settingsModal.js [hunk 3]
```diff
@@ -169,6 +196,18 @@ function SettingsModal() {
           />
         </Row>
       )}
+      {videoInputs.length > 0 && (
```
Comment (swissspidy) [inline packages/story-editor/src/components/mediaRecording/settingsModal.js:200]: Now that we have the quick action I think we can remove this from the settings modal again

Comment (merapi) [reply packages/story-editor/src/components/mediaRecording/settingsModal.js:200]: I asked about it on design sync, Andrew was in favor of leaving it, in the future I guess we will add other options like virtual background, etc.

The other thing that is worth considering is the fact that none of the quick actions are persisted - if you disable audio or video it will be enabled in the next session and so does the blur effect, but if you change it in the settings - it will be persisted (like other settings options), not sure if it's a bug or feature 😉

File: packages/story-editor/src/app/highlights/quickActions/useQuickActions.js [hunk 4]
```diff
@@ -843,6 +853,22 @@ const useQuickActions = () => {
...
       },
+      videoInput && {
+        Icon: videoEffect === 'blur' ? BackgroundBlur : BackgroundBlurOff,
+        label:
+          videoEffect === 'blur'
+            ? __('Disable Background Blur', 'web-stories')
+            : __('Enable Background Blur', 'web-stories'),
+        onClick: () => {
+          trackEvent('media_recording_background_blur', {
+            status: videoEffect === 'blur' ? 'off' : 'on',
```
Comment (swissspidy) [inline packages/story-editor/src/app/highlights/quickActions/useQuickActions.js]: Thinking: if we ever add more blur steps (e.g. strong/medium/light blur), then it might be easier to pass a number here, e.g.

```suggestion
            value: videoEffect === 'blur' ? 0 : 100,
```

File: packages/story-editor/src/components/mediaRecording/useTrim.js [hunk 5]
```diff
@@ -18,8 +18,12 @@
  * External dependencies
  */
 import { formatMsToHMS } from '@googleforcreators/media';
-import { useState, useCallback, useRef } from '@googleforcreators/react';
-import { useEffect } from 'react';
+import {
+  useState,
+  useCallback,
+  useRef,
+  useEffect,
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/useTrim.js:25]: **👏 praise:**

Good catch!

File: packages/story-editor/src/components/mediaRecording/settingsModal.js [hunk 6]
```diff
@@ -169,6 +196,18 @@ function SettingsModal() {
...
+        <Row>
+          <DropDown
+            ariaLabel={__('Video Effect', 'web-stories')}
+            placeholder={__('Select Video Effect', 'web-stories')}
+            options={videoEffects}
+            onMenuItemClick={onChangeVideoEffect}
+            selectedValue={localVideoEffect}
+            popupZIndex={11}
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/settingsModal.js]: **💬 suggestion:**
We have a `constants/zIndex.js` file, maybe add it to there instead?

File: packages/story-editor/src/components/mediaRecording/settingsModal.js [hunk 7]
```diff
@@ -107,6 +125,10 @@ function SettingsModal() {
   const audioInputs = mediaDevices
     .filter((device) => device.kind === 'audioinput')
     .map(({ deviceId, label }) => ({ value: deviceId, label }));
+  const videoEffects = [
+    { value: 'none', label: __('None', 'web-stories') },
+    { value: 'blur', label: __('Background blur', 'web-stories') },
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/settingsModal.js]: **🥜 nitpick:**
If we'll add other effects in the future, might be nice to create a constant for it. Could already now be `VIDEO_EFFECTS: { NONE: 'none', BLUR: 'blur' }` since `VIDEO.EFFECTS.BLUR` is already in multiple files.

File: packages/story-editor/src/components/mediaRecording/provider.js [hunk 8]
```diff
@@ -81,6 +83,9 @@ function MediaRecordingProvider({ children }) {
   const [audioInput, setAudioInput] = useState(
     localStore.getItemByKey(LOCAL_STORAGE_PREFIX.MEDIA_RECORDING_AUDIO_INPUT)
   );
+  const [videoEffect, setVideoEffect] = useState(
+    localStore.getItemByKey(LOCAL_STORAGE_PREFIX.MEDIA_RECORDING_VIDEO_EFFECT)
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/provider.js:87]: **💡 thought:**
Would it make sense to lazy init the state for performance reasons? Not sure how much of a difference it makes.

File: packages/story-editor/src/app/highlights/quickActions/useQuickActions.js [hunk 9]
```diff
@@ -331,6 +340,7 @@ const useQuickActions = () => {
     muteAudio: actions.muteAudio,
     unMuteAudio: actions.unMuteAudio,
     startTrim: actions.startTrim,
+    setVideoEffect: actions.setVideoEffect,
   }));
 
   const enableMediaRecordingTrimming = useFeature('recordingTrimming');
```
Comment (miina) [inline packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:351]: **💡 thought:**
Not related to this PR directly but should these actions be disabled during recording? Currently it seems that it's possible to click on it during recording which results in the recording just stopping. Thoughts?

Comment (miina) [reply packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:351]: Hm, okay, I'm now trying again and seeing these are disabled generally. Unfortunately I don't know how to reproduce it but somehow I ended up (twice) in a situation where I could click on the Disable Background Blur quick-action while recording -> recording stopped -> when I opened the recorder again, I saw this with neither of the buttons (Pause/Stop) doing anything:
Screenshot 2022-08-19 at 13 36 25

Comment (miina) [reply packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:351]: OK, I tried again and the following seems to break it eventually:
- Open the recorder
- Enable background blur
- Start recording
- Wait for a couple of seconds -- when testing, the quick-actions became enabled when the blur option was turned on.
- Now disable the blur while recording. And now the video stops recording. 
- Click X to close the recorder tool.

When I repeat the process a few times, the following occurs:
- If the video stopped recording because of using quick-actions during recording, and then when opening the recorder tool from the Library again, it starts recording right away when opening.
- After a few times of doing this, the video starts recording but doesn't really show the video and the Pause/Stop buttons don't do anything (as above on the screenshot).

**Review by miina — COMMENTED** [review]
(Will continue the review on Monday, couldn't go through everything)

File: packages/story-editor/src/app/highlights/quickActions/useQuickActions.js (hunk 4 above)
Comment (merapi) [reply packages/story-editor/src/app/highlights/quickActions/useQuickActions.js]: Do we want the historical data to be meaningful after the changes? If we do 0-100% of a scale, then the old data might be skewed if we change the scale. If we track blur px values (0px blur, 3px blur, 6px blur) then the old data will be still valid.

Comment (swissspidy) [reply packages/story-editor/src/app/highlights/quickActions/useQuickActions.js]: Good point with the scale 👍  Tracking the px values sounds more useful in that case

File: packages/story-editor/src/components/mediaRecording/provider.js (hunk 8 above)
Comment (merapi) [reply packages/story-editor/src/components/mediaRecording/provider.js:87]: We use this value as soon as you open the media recording screen not only in settings dialog.

File: packages/story-editor/src/app/highlights/quickActions/useQuickActions.js (hunk 9 above)
Comment (merapi) [reply packages/story-editor/src/app/highlights/quickActions/useQuickActions.js:351]: I thought it would be just a flag change (incorrect status/inputStatus/isReady), but it turned out that I was trying to fix something that was impossible to fix with the current version of the library. It was not possible to use the custom media recorder more than once per instance. Luckily, when I was about to go crazy I checked the library repo and saw [this one-liner](https://github.com/wmik/use-media-recorder/commit/94d06776129f7efdd9992785fcf2e2e400256a4d) 🪄 

_Note to self: always check if you are using the latest version of the library._

File: packages/story-editor/src/components/mediaRecording/playbackMedia.js [hunk 10]
```diff
@@ -40,31 +51,40 @@ function PlaybackMedia() {
...
     isProcessingTrim: state.isProcessingTrim,
+    setCanvasStream: actions.setCanvasStream,
+    setCanvasNode: actions.setCanvasNode,
   }));
   const setVideoNode = useVideoTrim(
     ({ actions: { setVideoNode } }) => setVideoNode
   );
 
   const isMuted = !hasAudio || isGif;
+  const hasVideoEffect = videoEffect && videoEffect !== 'none';
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/playbackMedia.js]:
```suggestion
  const hasVideoEffect = videoEffect && videoEffect !== VIDEO_EFFECTS.NONE;
```

File: packages/story-editor/src/components/mediaRecording/footer.js [hunk 11]
```diff
@@ -319,14 +329,20 @@ function Footer() {
   const { showSnackbar } = useSnackbar();
 
   const captureImage = useCallback(async () => {
-    if (!streamNode) {
+    const hasVideoEffect = videoEffect && videoEffect !== 'none';
```
Comment (miina) [inline packages/story-editor/src/components/mediaRecording/footer.js]:
```suggestion
    const hasVideoEffect = videoEffect && videoEffect !== VIDEO_EFFECTS.NONE;
```

**Review by miina — APPROVED** [review]
Works well now!

(Some unit tests are failing)

**Comment by felipebochehin87** [comment]
When switching on/off the blur background on Firefox, the image/video shakes/blinks/presents a weird behavior. Video below:


https://images.zenhubusercontent.com/235435637/dfc6390c-3c79-4894-b438-3c89660bf4f4/2022_09_06_10_58_55.mp4

**Comment by spacedmonkey** [comment]
Screenshot 2022-09-12 at 14 32 38
The dialog seems a little confusing. Why don't any of the dropdown have label / legend to explain what these are. Depending on the name of your webcam, it really might not be clear what is happening here.

**Comment by spacedmonkey** [comment]
Still not seeing this functionality working in safari. 
https://user-images.githubusercontent.com/237508/189669000-cdc04a62-a3f2-4c80-be5e-f0503eab0748.mov

**Comment by spacedmonkey** [comment]
Still seeing the permissions popup in firefox. 

https://user-images.githubusercontent.com/237508/189669720-7634e4f5-1f6d-46ef-92ed-a7def99ab61e.mov

**Comment by merapi** [comment]
> Still seeing the permissions popup in firefox.

What happens when you select the first camera again?
What happens after you refresh the page and select the first/second camera?

**Comment by merapi** [comment]
> The dialog seems a little confusing. Why don't any of the dropdown have label / legend to explain what these are. Depending on the name of your webcam, it really might not be clear what is happening here.

I guess there was no design/UX on this. I'm also not a fan of inputs without labels (or rather one-time labels, until you select something).

**Comment by merapi** [comment]
> Still not seeing this functionality working in safari.

Yep, Safari is tricky (when fetching frames from the canvas), still testing.

**Comment by spacedmonkey** [comment]
Noticed another weird issues with the settings dialog.  See video. 

https://user-images.githubusercontent.com/237508/189875958-87f31562-1bb6-4e20-bd56-6c2afd96c742.mov

**Comment by merapi** [comment]
> Noticed another weird issues with the settings dialog. See video.
> 
>  Screen.Recording.2022-09-13.at.11.14.03.mov

What browser is this? any errors in the console?

**Comment by spacedmonkey** [comment]
> What browser is this? any errors in the console?

Chrome 104. Nope.

**Comment by spacedmonkey** [comment]
Seeing slow down of image in safari. 
https://user-images.githubusercontent.com/237508/190190745-4cd586d1-1e16-4403-ad02-b3ae0e38bd96.mov

And other issues in firefox. 
https://user-images.githubusercontent.com/237508/190190940-63efdbe5-bb73-4582-9cd9-bd60a170b504.mov

**Comment by merapi** [comment]
> Seeing slow down of image in safari.

That is expected. It will never match the native blur or get close to it, tho it looks better on my Macbook Air.

> And other issues in firefox.

That is only happening on 2+ try in the same session, right?

**Comment by spacedmonkey** [comment]
Not seeing device list issue anymore.

**Comment by spacedmonkey** [comment]
Safari I am seeing this error when I try active blur background. 

Screenshot 2022-09-14 at 17 16 29
Screenshot 2022-09-14 at 17 15 57

**Comment by spacedmonkey** [comment]
Not seeing firefox issues now. 🎉

**Comment by merapi** [comment]
> Safari I am seeing this error when I try active blur background.

Yes, it's WASM and some of its memory issues. I added a try/catch there that's the best we can do.

**Comment by spacedmonkey** [comment]
> Yes, it's WASM and some of its memory issues. I added a try/catch there that's the best we can do.

More testing, I am seeing the slow version in safari. Blur background works but it is very slow in safari. 

https://user-images.githubusercontent.com/237508/190210854-10d83e0a-ca09-4916-bff4-c07e94a49a9f.mov
